broadcast/broadcast.go

```diff
@@ -0,0 +1,301 @@
+			default:
+				slot.cond.Wait()
+				slotSeq = slot.seq
+			}
+		}
+		msg := slot.data
+		slot.cond.L.Unlock()
+
+		var res EventOrLag[T]
+		if slotSeq == nextSeq+1 {
+			sub.nextSeq.Add(1)
+			res.event = msg
+		} else {
+			// Slot has been overwritten; compute the oldest sequence that is still
+			// available and inform the subscriber to skip ahead.
+			// newest and oldest are derived from the current tail. Because we only
+			// reach here when overwriting has happened, newest - bcast.capacity + 1 cannot
+			// underflow.
+			tail := bcast.tail.Load()
+			newest := tail - 1
+			oldest := newest - bcast.capacity + 1
+
+			res.lag.MissedSeq = nextSeq
+			res.lag.NextSeq = oldest
+			res.isLag = true
+
+			sub.nextSeq.Store(oldest)
+		}
+		// Deliver the result. If the downstream is slow (out is full), this blocks
+		// until either the subscriber reads or the subscription context is canceled.
+		select {
+		case sub.out <- res:
+			continue
+		case <-sub.done:
+			return
+		}
+	}
+}
+
+// Recv returns the user-facing channel for this subscription.
+func (sub *Subscription[T]) Recv() <-chan EventOrLag[T] {
+	return sub.out
+}
+
+// Unsubscribe terminates the subscribers delivery goroutine.
+func (sub *Subscription[T]) Unsubscribe() {
+	select {
+	case <-sub.done:
+		// Already closed
+	default:
+		close(sub.done)
+		// wake if subscriber is waiting
+		bcast := sub.bcast
+		idx := sub.nextSeq.Load() & bcast.mask
+		slot := &bcast.buffer[idx]
+		slot.cond.L.Lock()
+		slot.cond.Broadcast()
+		slot.cond.L.Unlock()
+	}
+}
+
+// A single ring buffer slot with a per-slot RWMutex.
+// - seq: published sequence number currently stored in this slot.
+// - mu: coordinates access to this slot.
+// - data: payload for the sequence in this slot.
+type slot[T any] struct {
+	seq  uint64
+	mu   sync.Mutex
+	cond sync.Cond
+	data T
+}
+
+// Broadcast is the fan-out ring buffer.
+// Fields:
+// - buffer:   ring buffer of slots.
+// - mu:       global mutex serialising all Send calls (MPMC-safe, single-writer).
+// - tail:     next sequence to assign (number of messages published so far).
+// - capacity: ring capacity (power of two, rounded up from the requested value).
+// - mask:     capacity-1 for index masking.
+// - ctx:      broadcast lifetime context; when canceled, Send returns ErrClosed.
+// - cancel:   function to cancel ctx
+type Broadcast[T any] struct {
+	buffer []slot[T]
+	mu     sync.Mutex
+	tail   atomic.Uint64
+
+	capacity uint64
+	mask     uint64
+
+	done chan struct{}
+}
+
+// New constructs a Broadcast with a ring of at least the given capacity.
+// - The actual capacity is rounded up to the next power of two.
+// - Each slot’s condition variable is initialised to use the slot’s mutex.
+// - The initial tail is 0 (no messages published).
+func New[T any](capacity uint64) *Broadcast[T] {
+	capacity = nextPowerOfTwo(capacity)
+	rb := make([]slot[T], capacity)
+	for i := range rb {
+		slot := &rb[i]
+		slot.cond = *sync.NewCond(&slot.mu)
```

Comment:
We're not using `mu` directly anywhere else, so probably we can remove `mu` from `slot` and also remove this initialization loop.

Reply:
I forgot that `L` is an interface, not a struct, so it doesn't have valid zero value.
We still need this loop, although I still think we can remove `slot.mu`